### PR TITLE
CORE-1912: websocketclient objects are not reuseable

### PR DESCRIPTION
**Pull request by harbu**
- IDEA file commits are intentional.
- Added test that reproduces the stack trace found on the latest comment of Jira ticket CORE-1912.
- Added the corresponding fix to method `publish`. When we are not connected, it will only `connect()` if `!userWantsToConnect`, i.e., we don't have an ongoing reconnect thread in the works.

### CODE REVIEW

File: src/main/java/com/streamr/client/StreamrClient.java [hunk 1]
```diff
@@ -447,7 +447,14 @@ public void publish(Stream stream, Map<String, Object> payload, Date timestamp,
 
     public void publish(Stream stream, Map<String, Object> payload, Date timestamp, String partitionKey, UnencryptedGroupKey newGroupKey) {
         if (!getState().equals(ReadyState.OPEN)) {
-            connect();
+            if (!userWantsToConnect) {
```
Comment (hpihkala) [inline src/main/java/com/streamr/client/StreamrClient.java:450]: **[minor]** This reads a bit funny due to current variable naming: "if the user doesn't want to connect, then connect!". I had to search through the rest of the file to figure out what `userWantsToConnect` actually is. 

However, I couldn't immediately come up with a clearly better name for the variable. A couple of (imperfect) suggestions that came to mind:
- `stayingConnected`
- `connectionLoopStarted`

Alternatively just a comment along the lines of "call connect() if the client isn't already attempting to stay connected" might do the trick.

File: src/test/groovy/com/streamr/client/StreamrClientSpec.groovy [hunk 2]
```diff
@@ -169,6 +169,42 @@ class StreamrClientSpec extends Specification {
...
+                        client.publish(stream, ["test": i])
+                    } catch (e) {
+                        errors.add(e)
+                    }
+                }
+            }.start()
+        }
+
+        then:
+        errors == []
```
Comment (hpihkala) [inline src/test/groovy/com/streamr/client/StreamrClientSpec.groovy:204]: Should wait for all the threads to finish before checking `errors`. Now there's a possible execution path, where `errors == []` is checked before any of the threads have made it to `publish` yet.

Something along the lines of:

```
then:
new PollingConditions().within(10) {
    publishThreads.find { it.isAlive() } == null
}
errors == []
```

File: src/main/java/com/streamr/client/StreamrClient.java [hunk 3]
```diff
@@ -447,7 +447,14 @@ public void publish(Stream stream, Map<String, Object> payload, Date timestamp,
 
     public void publish(Stream stream, Map<String, Object> payload, Date timestamp, String partitionKey, UnencryptedGroupKey newGroupKey) {
```
Comment (hpihkala) [inline src/main/java/com/streamr/client/StreamrClient.java:448]: Should we simply add `synchronized` keyword to `publish`? That would secure the rest of the method from race conditions (access to `keyExchangeUtil` on lines 463-465 for example).

File: src/test/groovy/com/streamr/client/StreamrClientSpec.groovy (hunk 2 above)
Comment (harbu) [reply src/test/groovy/com/streamr/client/StreamrClientSpec.groovy:204]: Good catch

File: src/main/java/com/streamr/client/StreamrClient.java (hunk 1 above)
Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java:450]: So the thinking here is similar to that of `autoConnect = true` in the streamr-client-javascript library: a convenience feature for the user to not have to invoke `connect` before `publish`. But yeah I definitely see how the naming is quite confusing here.  I'm leaning on renaming to `stayingConnected` _and_ adding a clarifying comment.

File: src/main/java/com/streamr/client/StreamrClient.java (hunk 3 above)
Comment (harbu) [reply src/main/java/com/streamr/client/StreamrClient.java:447]: As discussed in Slack. let's leave this out for now and re-consider later.

**Pull request merged**